src/Mvc/Mvc.Core/src/ModelBinding/FormFileValueProvider.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding
+{
+    /// <summary>
+    /// An <see cref="IValueProvider"/> adapter for data stored in an <see cref="IFormFileCollection"/>.
+    /// </summary>
+    /// <remarks>
+    /// Unlike most <see cref="IValueProvider"/> instances, <see cref="FormFileValueProvider"/> does not provide any values, but
+    /// specifically responds to <see cref="ContainsPrefix(string)"/> queries. This allows the model binding system to
+    /// recurse in to deeply nested object graphs with only values for form files.
+    /// </remarks>
+    public sealed class FormFileValueProvider : IValueProvider
```

Comment:
What motivated the choice to make this separate from `FormValueProvider`? Is the idea that someone could remove this value provider if they don't want it?

Reply:
Yeah, that was one of the reasons. The other reason is that `FormValueProvider` is meant to be derived, and it now makes user code much more complicated since they now have to track two separate prefix groups.